### PR TITLE
CLOUDP-32539 POST stats in an idempotent way.

### PR DESCRIPTION
**Pull request by chunmingli**
https://jira.mongodb.org/browse/CLOUDP-32539

This patch includes:
* POST stats counters as compared to deltas
* Failed(), Succeeded() and InProgress() no longer push Progress synchronously
* Made `StatsPost` an anonymous struct
* Auth key now becomes a path param and is removed from the stats struct. Previously it's a form param.
* More verbose logging

Java side PR to come separately.

### CODE REVIEW

File: status.go [hunk 1]
```diff
@@ -161,52 +167,52 @@ func (p *StatsTracker) Failed(req *RequestStat) {
...
 }
 
 func (p *DownloadStats) addRequestToMap(req *RequestStat) {
 	p.Lock()
-	if p.requestStatMap == nil {
-		p.requestStatMap = make(map[uint64]*RequestStat)
-	}
-	p.Unlock()
-
+	defer p.Unlock()
```
Comment (lawctan) [inline status.go]: Will this work? p.removeOldestFailedRequest() already does locking. I believe Go's sync.mutex doesn't allow re-entrant locks

File: status.go [hunk 2]
```diff
@@ -81,19 +77,31 @@ func NewStatsTracker(centralUrl, jobId, authKey string) *StatsTracker {
...
+func (p *StatsTracker) Push() {
+	log.Printf("Prepare to push status: %v", p.stats)
+
 	if p.centralUrl == "" || p.jobId == "" {
-		log.Println("No CentralUrl or RestoreJobId. Running in local debug mode?")
+		log.Println("Cannot publish stats due to no CentralUrl or RestoreJobId. Running in local debug mode?")
 		return
 	}
-	body, err := json.Marshal(&StatsPost{inProgress, failed, succeeded, p.stats.TotalBytes, p.stats.getRequestStatsForPush(), p.authKey})
+	body, err := json.Marshal(&struct {
```
Comment (lawctan) [inline status.go]: what's the motivation for doing an anonymous struct vs explicit struct? Just wanna know so we can keep it in mind in the future

Comment (chunmingli) [reply status.go]: This is more of a personal preference, but for types/methods that are used only once, unless it _really_ helps with readability, I find easier when the type is defined inline so I don't need to jump/scroll around to see where the named explicit struct is.

File: status.go (hunk 1 above)
Comment (chunmingli) [reply status.go]: Good call. Testing it. FWIW removeOldestFailedRequest() also calls into getAllIdsFromMapInOrder() that locks the mutex as well.

Comment (chunmingli) [reply status.go]: I changed it such that when the method receiver is `ReqMap` (alias as `map[uint64]*RequestStat`) then no locking is in the method. Specifically, `func (requests ReqMap) getAllIdsSorted() []uint64 ` and `func (requests ReqMap) removeOldestInactiveRequest()` no longer has the locking logic. The caller needs to hold proper locks when calling into these two.

**Comment by chunmingli** [comment]
@lawctan ready for another look

File: status.go [hunk 3]
```diff
@@ -188,31 +185,29 @@ func (p *DownloadStats) addRequestToMap(req *RequestStat) {
...
-func (p *DownloadStats) removeOldestFailedRequest() {
-	keys := p.getAllIdsFromMapInOrder()
+func (requests ReqMap) removeOldestInactiveRequest() {
+	keys := requests.getAllIdsSorted()
 	for _, k := range keys {
-		req, _ := p.getRequest(k)
-		if req.Status == FailedString {
-			delete(p.RequestStats, k)
+		req, _ := requests[k]
+		if req.Status != InProgressString {
```
Comment (lawctan) [inline status.go]: we don't want to delete SuccessString either. So, checking for failed should be better.

File: atlas_restore.go [hunk 4]
```diff
@@ -191,7 +191,7 @@ func startProgressPushes(p *StatsTracker) {
 	ticker := time.NewTicker(time.Second * time.Duration(PushTimeInSeconds))
 	go func() {
 		for _ = range ticker.C {
-			p.PushProgress()
+			p.Push()
```
Comment (chunmingli) [inline atlas_restore.go]: I noticed that `p.PushProgress()` previously created a separate goroutine to do `go p.Push()`, which is unnecessary as the stats publishing goroutine should suffice.

**Comment by chunmingli** [comment]
Thanks for reviewing!

**Pull request merged**